tango/integrations/torch/eval.py

```diff
@@ -0,0 +1,187 @@
+            This may not be the correct behavior for some metrics (such as F1),
+            in which you should set this to ``False`` and handle the aggregation
+            internally in your model or with an :class:`EvalCallback`
+            (using :meth:`EvalCallback.post_batch()`).
+        callbacks : ``List[TrainCallback]``
+            A list of :class:`EvalCallback`.
+
+        """
+        set_seed_all(seed)
+
+        check_dataset(dataset_dict, test_split)
+
+        # Resolve device.
+        device = resolve_device()
+
+        # Prep model.
+        model = model.eval().to(device)
+
+        # Construct dataloader.
+        dataloader: DataLoader = dataloader.construct(dataset=dataset_dict[test_split])
+
+        steps: int
+        try:
+            dataloader_len = len(dataloader)
+            steps = dataloader_len if eval_steps is None else min(dataloader_len, eval_steps)
+        except TypeError:
+            if eval_steps is None:
+                raise ConfigurationError(
+                    "You must set 'eval_steps' for streaming/iterable datasets"
+                )
+            else:
+                steps = eval_steps
+
+        # Initialize callbacks.
+        callbacks: List[EvalCallback] = [
+            callback.construct(
+                step_id=self.unique_id,
+                work_dir=self.work_dir,
+                model=model,
+                dataset_dict=dataset_dict,
+                dataloader=dataloader,
+            )
+            for callback in (callbacks or [])
+        ]
+        for callback in callbacks:
+            callback.pre_eval_loop()
+
+        eval_batches = enumerate(islice(dataloader, steps))
+
+        running_metrics: Dict[str, float] = defaultdict(float)
+        aggregated_metrics: Dict[str, float] = {}
+
+        with Tqdm.tqdm(eval_batches, desc="Evaluating", total=steps) as batch_iter:
+            for step, batch in batch_iter:
+                should_log_this_step = step % log_every == 0 or step == steps - 1
+
+                for callback in callbacks:
+                    callback.pre_batch(step, batch)
+
+                batch = move_to_device(batch, device)
+                with torch.inference_mode():
+                    outputs = model(**batch)
+
+                for callback in callbacks:
+                    callback.post_batch(step, outputs)
+
+                # Gather metrics we want to track.
+                batch_metrics = {
+                    k: outputs[k].item() if isinstance(outputs[k], torch.Tensor) else outputs[k]
+                    for k in metric_names
+                }
+
+                # Aggregate metrics.
+                if auto_aggregate_metrics:
+                    for k in batch_metrics:
+                        running_metrics[k] += batch_metrics[k]
+                        aggregated_metrics[k] = running_metrics[k] / (step + 1)
+                else:
+                    aggregated_metrics.update(batch_metrics)
```

Comment:
I'm pretty excited about torchmetrics' latest release. It is not the right time to integrate that here now, but torchmetrics would make this go away (because all metrics already know how to aggregate). This works even across GPUs, which does not work here.